### PR TITLE
Rancher token provider

### PR DESCRIPTION
**Pull request by dmrogers7**
Adds Rancher as a token provider

The `provider` query parameter has been added to the `/tokens` endpoint and has two values
1. google
2. rancher
If no `provider` is supplied, then a google token is returned

### CODE REVIEW

File: pkg/http/token.go [hunk 1]
```diff
@@ -1,28 +1,48 @@
...
+		GetRancherToken(c)
+	case "google", "":
+		GetGoogleToken(c)
+	default:
+		c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("Unsupported token provider: %s", provider)})
+		return
+	}
+}
+
+func GetGoogleToken(c *gin.Context) {
```
Comment (billiford) [inline pkg/http/token.go:38]: We should make these unexported if possible.

```go
getGoogleToken(...)
getRancherToken(...)
```

File: pkg/http/token.go [hunk 2]
```diff
@@ -1,28 +1,48 @@
...
+	googleMux.Lock()
+	defer googleMux.Unlock()
 
 	if time.Since(t) > expiration || token == "" {
-		token, err = google.NewToken()
+		googleClient := google.Instance(c)
+		token, err = googleClient.NewToken()
+
 		if err != nil {
```
Comment (billiford) [inline pkg/http/token.go:46]: Let's format these lines to put the error next to the error check.
```go
		googleClient := google.Instance(c)
		
		token, err = googleClient.NewToken()
		if err != nil {
```

File: README.md [hunk 3]
```diff
@@ -16,11 +42,29 @@ go build cmd/arcade/arcade.go
 
 ```bash
 export ARCADE_API_KEY=test
+export RANCHER_ENABLED=TRUE
+export RANCHER_URL=https://rancher-np-ssc.homedepot.com/v3-public/activeDirectoryProviders/activedirectory?action=login
```
Comment (billiford) [inline README.md:46]: Let's do a more generic url...
```sh
export RANCHER_URL=https://rancher.example.com/v3-public/activeDirectoryProviders/activedirectory?action=login
```

File: cmd/arcade/arcade.go [hunk 4]
```diff
@@ -5,27 +5,54 @@ import (
 	"os"
 
 	"github.com/gin-gonic/gin"
+	"github.homedepot.com/cd/arcade/pkg/google"
```
Comment (billiford) [inline cmd/arcade/arcade.go:8]: I definitely started this issue... but why are we referencing `github.homedepot.com/cd/...`? Should we move this repo to the github org?

File: pkg/http/token_test.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.homedepot.com/cd/arcade/pkg/google/googlefakes"
+	arcade_http "github.homedepot.com/cd/arcade/pkg/http"
```
Comment (billiford) [inline pkg/http/token_test.go:16]: I can't find the source on this right now, but I don't think we should be putting underscores on our imported packages. Instead maybe just say `arcadehttp`.

File: pkg/rancher/payload_test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  "id": "kubeconfig-u-i76rfanbw5",
+  "isDerived": true,
+  "labels": {
+    "authn.management.cattle.io/kind": "kubeconfig",
+    "authn.management.cattle.io/token-userId": "u-i76rfanbw5",
+    "cattle.io/creator": "norman"
+  },
+  "lastUpdateTime": "",
+  "links": {
+    "self": "https://rancher-np-ssc.homedepot.com/v3-public/tokens/kubeconfig-u-i76rfanbw5"
```
Comment (billiford) [inline pkg/rancher/payload_test.go:25]: `https://rancher.example.com/v3-public/tokens/kubeconfig-u-i76rfanbw5`

**Review by billiford — CHANGES_REQUESTED** [review]
I think I messed up the module name when I originally set this up, as everything is referencing `github.homedepot.com/cd/...`. Should we consider moving this to the open-source homedepot org?

**Comment by billiford** [comment]
@dmrogers7 We can address the module issue outside of this PR unless you've already fixed it.

File: pkg/http/token_test.go (hunk 5 above)
Comment (billiford) [reply pkg/http/token_test.go:16]: Well I found the source that is referencing package names. I usually follow this convention for import names as well. https://blog.golang.org/package-names

File: README.md [hunk 7]
```diff
@@ -16,11 +42,29 @@ go build cmd/arcade/arcade.go
...
+export RANCHER_PASSWORD=myPassword
 ./arcade
 ```
 
 ### Test
 
+**Google**
+
+```bash
+curl localhost:1982/tokens/provider=google -H "Api-Key: test"
```
[the remark this answers is not in the document]
Comment (dmrogers7) [reply README.md]: doh!

**Comment by dmrogers7** [comment]
I made the fixes, including changing it to `gibhub.com/homedepot`.  Once this PR is merged, then we can fork the repo into the homedepot org

**Review by billiford — APPROVED** [review]
LGTM!

**Pull request merged**